src/light-clients/ethereum.md

```diff
@@ -0,0 +1,143 @@
+1. Updates the client's internal state to reflect the new "latest verified block"
+
+Updates can happen in sequence (verifying each block) or can skip intermediate blocks using more complex proof mechanisms. The efficiency of this process is what makes light clients practical for cross-chain communication.
+
+If the light client detects conflicting information or invalid proofs that suggest an attack attempt, it can enter a "frozen" state. This is a safety mechanism that prevents the client from processing potentially fraudulent updates. Recovery from a frozen state typically requires governance intervention.
+
+Since initialization is rather trivial, we will not dive deeper into it.
+
+### Updating
+
+Since Ethereum is finalized by the beaconchain, our Ethereum light client accepts beacon block data as update input. A [beacon block](https://eth2book.info/capella/part3/containers/blocks/#beacon-blocks) rougly has this structure:
+
+```python
+class BeaconBlockBody(Container):
+    randao_reveal: BLSSignature
+    eth1_data: Eth1Data
+    graffiti: Bytes32
+    proposer_slashings: List[ProposerSlashing, MAX_PROPOSER_SLASHINGS]
+    attester_slashings: List[AttesterSlashing, MAX_ATTESTER_SLASHINGS]
+    attestations: List[Attestation, MAX_ATTESTATIONS]
+    deposits: List[Deposit, MAX_DEPOSITS]
+    voluntary_exits: List[SignedVoluntaryExit, MAX_VOLUNTARY_EXITS]
+    sync_aggregate: SyncAggregate
+    execution_payload: ExecutionPayload
+    bls_to_execution_changes: List[SignedBLSToExecutionChange, MAX_BLS_TO_EXECUTION_CHANGES]
+```
+
+We are specifically interested in `sync_aggregate`, which is a structure describing the votes of the sync committee:
+
+```python
+class SyncAggregate(Container):
+    sync_committee_bits: Bitvector[SYNC_COMMITTEE_SIZE]
+    sync_committee_signature: BLSSignature
+```
+
+The `sync_committee_bits` indicate which members voted (not all need to vote), and the `sync_committee_signature` is a BLS signature of the members references in the bit vector.
+
+BLS signatures (Boneh-Lynn-Shacham) are a type of cryptographic signature scheme that allows multiple signatures to be aggregated into a single signature. This makes them space and compute efficient (you can aggregate hundreds of signatures into one). Just as we aggregate signatures, we can aggregate public keys as well, such that the aggregate public key can verify the aggregated signature.
+
+For our SyncAggregate, computing the aggregate pubkey is simple:
+
+```python
+def _aggregate_pubkeys(committee, bits)
+    pubkeys = []
+    for i, bit in enumerate(bits):
+        if bit:
+            pubkeys.append(committee[i])
+    return bls.Aggregate(pubkeys)
+```
+
+At scale, we can aggregate thousands (if not hundreds of thousands) of signatures and public keys, whily only verifying their aggregates.
+
+To our light client, as long as a majority of sync committee members have attested the block, it is consider final.
```

Comment:
I'm not sure if we want to talk about the optimization that we did here. But right now, since it's guaranteed that the majority of the sync committee will sign, instead of aggregating their public keys, we aggregate the additive inverse of the nonsigners' public keys and add it to the aggregate public key. We calculate and save the additive inverse public keys during every Sync committee period changing update.